numba/tests/test_listobject.py

```diff
@@ -935,6 +935,17 @@ def foo():
             str(raises.exception),
         )
 
+    def test_list_extend_unicode_type(self):
```

Comment:
This test doesn't test any type refinement, since the list is being initialized with the correct type already. Can you check if the test passes also without the changes made above.

Reply:
It does indeed pass without the changes. Will I need to create a List without initialization? How do I do that?

Reply:
Does this look alright?

```python
    def test_list_extend_unicode_type(self):
        @njit
        def foo(string):
            l = List()
            l.extend(string)
            return len(l)

        self.assertEqual(foo(""), 0)
        self.assertEqual(foo("abc"), 3)
        self.assertEqual(foo("\nabc\t"), 5)
```

Reply:
So that is a good start, these are the things I would change:

a) You'll probably want to move this test to the `tests/test_typedlist.py`, there is a section related to `extend` there already.

b) You'll want to test both the `foo`, but also the `foo.py_func` to make sure this works fine outside of an `@njit` context.

c) Lastly, I would recommend to write an ancillary function that uses `append` (assuming that `append` works) to create an expected result, rather than just checking the length. I.e. something that goes a bit deeper than just the length.

Reply:
Almost done with this except I ran into an issue where `foo.py_func` throws an error with string null input `""`.

```
======================================================================
ERROR: test_list_extend_unicode_type (numba.tests.test_typedlist.TestTypedList)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/rohitsanjay/numba/numba/tests/test_typedlist.py", line 117, in test_list_extend_unicode_type
    self.assertEqual(list(func("")), [])
  File "/Users/rohitsanjay/numba/numba/tests/test_typedlist.py", line 113, in foo
    l.extend(string)
  File "/Users/rohitsanjay/numba/numba/typed/typedlist.py", line 323, in extend
    self._initialise_list(iterable[0])
IndexError: string index out of range

----------------------------------------------------------------------
```
I guess this is a known issue, looking at the FIXME - https://github.com/numba/numba/blob/master/numba/typed/typedlist.py#L318-L326